crates/adapters/src/controller/mod.rs

```diff
@@ -413,13 +416,25 @@ impl Controller {
     ) -> Result<(), ControllerError>
     where
         F: FnOnce(
-            usize,
+            CircuitConfig,
         )
             -> Result<(Box<dyn DbspCircuitHandle>, Box<dyn CircuitCatalog>), ControllerError>,
     {
         let mut start: Option<Instant> = None;
 
-        let mut circuit = match circuit_factory(controller.status.global_config.workers as usize) {
+        let config = CircuitConfig {
+            layout: Layout::new_solo(controller.status.global_config.workers as usize),
+            storage: controller
+                .server_args
+                .storage_location
+                .clone()
+                .unwrap_or_else(|| {
+                    tempfile::tempdir()
+                        .map(|dir| dir.into_path())
+                        .expect("Can't create temporary storage")
+                }),
```

Comment:
This is not an ideal use of `tempdir` because using `into_path()` means that it will not be deleted automatically by the destructor, so we'll end up with stray temporary directories on each run. (See https://docs.rs/tempfile/latest/tempfile/struct.TempDir.html#resource-leaking).

Reply:
Hmm, maybe it's intentional? We want the directory to be persistent sometimes. But we'll still get lots of temporary directories if we're testing, I think.

Reply:
Yes, for now this was intentional as it was beneficial to be able to inspect the state of the files once the program/tests have exited. Maybe I can change it to preserve it if things go wrong/panic and delete it otherwise. In any case probably needs a separate mechanism from `tempfile`

Reply:
I reworked this so that temporary storage location is always removed after the runtime is destroyed (except when we panic).